flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1697,133 +1621,620 @@ class ExprLowering {
+  }
+  CC genarr(const Fortran::evaluate::StructureConstructor &x) {
+    TODO("structure constructor");
+    return [](IterSpace iters) -> ExtValue { return mlir::Value{}; };
+  }
+  template <int KIND>
+  CC genarr(const Fortran::evaluate::Not<KIND> &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
+  }
+  template <int KIND>
+  CC genarr(const Fortran::evaluate::LogicalOperation<KIND> &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
+  }
+  CC genarr(
+      const Fortran::evaluate::Relational<Fortran::evaluate::SomeType> &x) {
+    return [](IterSpace iters) -> ExtValue {
+      return mlir::Value{}; /* FIXME */
+    };
   }
 
 private:
+  explicit ArrayExprLowering(Fortran::lower::AbstractConverter &converter,
+                             Fortran::lower::StatementContext &stmtCtx,
+                             Fortran::lower::SymMap &symMap,
+                             fir::ArrayLoadOp dst = {})
+      : converter{converter}, builder{converter.getFirOpBuilder()},
+        stmtCtx{stmtCtx}, symMap{symMap}, destination{dst} {}
+
+  explicit ArrayExprLowering(Fortran::lower::AbstractConverter &converter,
+                             Fortran::lower::StatementContext &stmtCtx,
+                             Fortran::lower::SymMap &symMap, bool projection,
+                             fir::ArrayLoadOp dst = {})
+      : converter{converter}, builder{converter.getFirOpBuilder()},
+        stmtCtx{stmtCtx}, symMap{symMap}, destination{dst}, inProjection{
+                                                                projection} {}
+
   mlir::Location getLoc() { return converter.getCurrentLocation(); }
+  void setInProjection(bool on = true) { inProjection = on; }
 
   Fortran::lower::AbstractConverter &converter;
   Fortran::lower::FirOpBuilder &builder;
+  Fortran::lower::StatementContext &stmtCtx;
   Fortran::lower::SymMap &symMap;
-  std::vector<mlir::Value> loopCounters;
-  std::vector<mlir::Value> arrayBlockArgs;
-  std::vector<mlir::Value> loopReturnVals;
-  const Fortran::lower::ExpressionContext &exprCtx;
+  fir::ArrayLoadOp destination;
+  bool inProjection{false};
 };
 } // namespace
 
-fir::ShapeOp Fortran::lower::ExpressionContext::getShape() const {
-  // Verify that this context is built correctly.
-  if (auto sh = mlir::dyn_cast<fir::ShapeOp>(shape.getDefiningOp()))
-    return sh;
-  llvm::report_fatal_error("operation must be a fir.shape");
-}
+namespace {
+// FIXME: The `AnyTraverse` pattern doesn't short-circuit its visitation. That
+// is, the entire `Ev::Expr` tree is visited, but only the first non-`None`
+// result wins. The root cause is the use of `Combine()` in the `Traverse`
+// visitor pattern. Furthermore, the `operator()` methods are all `const`, a
+// usage obstacle such that any visitor class has to use `mutable` internal
+// state to be able to carry its own context.
+struct IntraexprShapeAnalyzer
```

Comment:
Should it be called `IntraExprShapeAnalyzer` ?
The class may deserve a quick description of what it is doing on top of the rational of the AnyTraverse issue description.

Reply:
L'anglais est une langue étrange, non?